File: bindings/java-jni/src/main/java/org/maplibre/nativejni/query/FeatureExtensionResult.java
package org.maplibre.nativejni.query;

import java.util.List;
import java.util.Objects;
import org.maplibre.nativejni.geo.Feature;
import org.maplibre.nativejni.json.JsonValue;

/** Copied result from a feature extension query. */
public sealed interface FeatureExtensionResult
    permits FeatureExtensionResult.Value,
        FeatureExtensionResult.FeatureCollection,
        FeatureExtensionResult.Unknown {
  record Value(JsonValue value) implements FeatureExtensionResult {
    public Value {
      Objects.requireNonNull(value, "value");
    }
  }

  record FeatureCollection(List<Feature> features) implements FeatureExtensionResult {
    public FeatureCollection {
      features = List.copyOf(features);
    }
  }

  record Unknown(int rawType) implements FeatureExtensionResult {}
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/query/FeatureExtensionResult.kt
package org.maplibre.nativeffi.query

import org.maplibre.nativeffi.geo.Feature
import org.maplibre.nativeffi.json.JsonValue

/** Copied result from a feature extension query. */
public sealed interface FeatureExtensionResult {
  public data class Value(public val value: JsonValue) : FeatureExtensionResult

  public class FeatureCollection(features: List<Feature>) : FeatureExtensionResult {
    public val features: List<Feature> = features.toList()

    override fun equals(other: Any?): Boolean =
      other is FeatureCollection && features == other.features

    override fun hashCode(): Int = features.hashCode()

    override fun toString(): String = "FeatureCollection(features=$features)"
  }

  public data class Unknown(public val rawType: Int) : FeatureExtensionResult
}
